### PR TITLE
Custom Task Tutorial

### PR DESCRIPTION
**Pull request by teddykoker**
TODO

### CODE REVIEW

File: docs/source/custom_task.rst [hunk 1]
```diff
@@ -1,37 +1,139 @@
...
-You can create your own Flash task to if you want to solve any other deep learning problem.
+In this tutorial we will go over the process of creating a custom task,
+along with a custom data module.
 
-See this example for defining a linear classifier task:
-
-.. testcode::
+.. code:: ipython3
 
+    import flash
```
Comment (aribornstein) [inline docs/source/custom_task.rst:9]: You should introduce that you are importing the necessary dependencies. Why do we need sklearn?

File: docs/source/custom_task.rst [hunk 2]
```diff
@@ -1,37 +1,139 @@
...
-            super().__init__(model=None,
+    from torch.utils.data import TensorDataset, DataLoader
+    from torch import nn
+    from sklearn import datasets
+    from sklearn.model_selection import train_test_split
+
+The Task
+--------
+
+Here we create a basic linear regression task by subclassing
```
Comment (aribornstein) [inline docs/source/custom_task.rst:20]: Put a link to explaining what linear regression is

File: docs/source/custom_task.rst [hunk 3]
```diff
@@ -1,37 +1,139 @@
...
                 metrics=metrics,
                 learning_rate=learning_rate,
             )
-            self.save_hyperparameters()
+            
+        def forward(self, x):
+            return self.model(x)
 
-            self.linear = torch.nn.Linear(num_inputs, num_classes)
+Where is the training step?
```
Comment (aribornstein) [inline docs/source/custom_task.rst:48]: Where **are** the training step?

File: docs/source/custom_task.rst [hunk 4]
```diff
@@ -1,37 +1,139 @@
...
+Because of our custom data pipeline’s ``after_uncollate`` method, we
+will get a nicely formatted output like the following:
+
+::
+
+   ['disease progression: 155.90',
+    'disease progression: 156.59',
+    'disease progression: 152.69',
+    'disease progression: 149.05',
+    'disease progression: 150.90']
```
Comment (tchaton) [inline docs/source/custom_task.rst:139]: Let's add the DataPipeline object to the doc.

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/custom_task.rst [hunk 5]
```diff
@@ -1,37 +1,139 @@
...
 
-            self.linear = torch.nn.Linear(num_inputs, num_classes)
+Where is the training step?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-        def forward(self, x):
-            return self.linear(x)
+Most models can be trained simply by passing the output of ``forward``
+to the supplied ``loss_fn``, and then passing the resulting loss to the
+supplied ``optimizer``. If you need a more custom configuration, you can
```
Comment (aribornstein) [inline docs/source/custom_task.rst:53]: This is not so accessible. Make it one level more accessible

**Review by aribornstein — CHANGES_REQUESTED** [review]
This is pretty clear if you know lightning but I don't think it's accesible to a flash audience. I think many people will have trouble extrapolating this process to their own models  if you want to sit together I can help you make the tone and content more accesible.

File: docs/source/custom_task.rst [hunk 6]
```diff
@@ -1,37 +1,139 @@
...
-            metrics=[Accuracy()],
-            learning_rate=1e-3,
-        ):
-            super().__init__(model=None,
+    from torch.utils.data import TensorDataset, DataLoader
+    from torch import nn
+    from sklearn import datasets
+    from sklearn.model_selection import train_test_split
+
+The Task
```
Comment (carmocca) [inline docs/source/custom_task.rst:17]:
```suggestion
The Task: Linear regression
```

File: docs/source/custom_task.rst [hunk 7]
```diff
@@ -1,37 +1,139 @@
...
+--------
+
+Here we create a basic linear regression task by subclassing
+``flash.Task``. For the majority of tasks, you will likely only need to
+override the ``__init__`` and ``__forward__`` methods of task.
```
Comment (carmocca) [inline docs/source/custom_task.rst:22]:
```suggestion
override the ``__init__`` and ``__forward__`` methods.
```

File: docs/source/custom_task.rst [hunk 8]
```diff
@@ -1,37 +1,139 @@
...
+                model=model,
                 loss_fn=loss_fn,
                 optimizer=optimizer,
                 metrics=metrics,
                 learning_rate=learning_rate,
             )
-            self.save_hyperparameters()
+            
+        def forward(self, x):
+            return self.model(x)
```
Comment (carmocca) [inline docs/source/custom_task.rst:46]:
```suggestion
            # we don't actually need to override this method for this example
            return self.model(x)
```

File: docs/source/custom_task.rst (hunk 3 above)
Comment (carmocca) [reply docs/source/custom_task.rst:48]: Where are the training/validation/test steps?

File: docs/source/custom_task.rst [hunk 9]
```diff
@@ -1,37 +1,139 @@
...
+
+For a task you will likely need a specific way of loading data. For this
+example, lets say we want a ``flash.DataModule`` to be used explicitly
+for the prediction of diabetes disease progression. We can create this
+``DataModule`` below, wrapping the scikit-learn `Diabetes
+dataset <https://scikit-learn.org/stable/datasets/toy_dataset.html#diabetes-dataset>`__.
+
+.. code:: ipython3
+
+    class DiabetesPipeline(flash.core.data.TaskDataPipeline):
```
Comment (carmocca) [inline docs/source/custom_task.rst:71]: I guess we should introduce the pipeline before mentioning it here

**Comment by edenlightning** [comment]
I suggest something in this pattern: for each object and the API explain what should be added, returned, and why.

The Task
========

To create a custom task you will need:

1. A model- this can be an :class:`~torch.nn.module` or a :class:`~pytorch_lightning.LightningModule`.

2. A loss function

3. An optimizer.

To create the task, subclass the :class:`~flash.core.model.Task`. For the majority of tasks, you will likely only need to
override the :func:`~flash.core.model.Task.__init__` and :func:`~flash.core.model.Task.forward` methods.


* The :func:`~flash.core.model.Task.__init__` method should set the model, the optimizer and the loss function you want to use in your task, and pass them to `super`.

.. code:: ipython3

    class LinearRegression(flash.Task):
        def __init__(self, num_inputs, learning_rate=0.001, metrics=None):

            # what kind of model do we want?
            model = nn.Linear(num_inputs, 1)
    
            # what loss function do we want?
            loss_fn = torch.nn.functional.mse_loss
            
            # what optimizer to do we want?
            optimizer = torch.optim.SGD
            
            super().__init__(
                model=model,
                loss_fn=loss_fn,
                optimizer=optimizer,
                metrics=metrics,
                learning_rate=learning_rate,
            )

* The step method.....


.. code:: ipython3


* The :func:`~flash.core.model.Task.forward` method will define the operations you want to use for prediction.

.. code:: ipython3

    #TODO: add a better example here
....


* The predict method ...


* Data pipeline...
